pkg/clients/eks/eks.go

```diff
@@ -19,702 +19,338 @@ package eks
+		if len(in.ResourcesVpcConfig.PublicAccessCidrs) == 0 && len(cluster.ResourcesVpcConfig.PublicAccessCidrs) > 0 {
+			in.ResourcesVpcConfig.PublicAccessCidrs = cluster.ResourcesVpcConfig.PublicAccessCidrs
+		}
 
-	request := e.amiClient.DescribeImagesRequest(&ec2.DescribeImagesInput{
-		Filters: ec2Filters,
-	})
-	out, err := request.Send(context.TODO())
-
-	if err != nil {
-		return nil, err
+		if len(in.ResourcesVpcConfig.SecurityGroupIDs) == 0 && len(cluster.ResourcesVpcConfig.SecurityGroupIds) > 0 {
+			in.ResourcesVpcConfig.SecurityGroupIDs = cluster.ResourcesVpcConfig.SecurityGroupIds
+		}
+		if len(in.ResourcesVpcConfig.SubnetIDs) == 0 && len(cluster.ResourcesVpcConfig.SubnetIds) > 0 {
+			in.ResourcesVpcConfig.SubnetIDs = cluster.ResourcesVpcConfig.SubnetIds
+		}
 	}
+	in.RoleArn = awsclients.LateInitializeStringPtr(in.RoleArn, cluster.RoleArn)
+	in.Version = awsclients.LateInitializeStringPtr(in.Version, cluster.Version)
+}
 
-	result := make([]*ec2.Image, len(out.Images))
-	for i := 0; i < len(out.Images); i++ {
-		result[i] = &out.Images[i]
+// IsUpToDate checks whether there is a change in any of the modifiable fields.
+func IsUpToDate(p *v1beta1.ClusterParameters, cluster *eks.Cluster) (bool, error) {
+	patch, err := CreatePatch(cluster, p)
+	if err != nil {
+		return false, err
 	}
 
-	return result, nil
-}
-
-func getMostRecentImage(images []*ec2.Image) *ec2.Image {
-	var result *ec2.Image
-	for i := 0; i < len(images); i++ {
-		if result == nil || aws.StringValue(result.CreationDate) < aws.StringValue(images[i].CreationDate) {
-			result = images[i]
+	// NOTE(hasheddan): AWS removes insignificant bits from CIDRs, so we must
+	// compare by converting user-supplied CIDRs to network blocks. We only skip
+	// comparison if both external and local have no CIDR blocks defined.
+	if (cluster.ResourcesVpcConfig != nil && len(cluster.ResourcesVpcConfig.PublicAccessCidrs) > 0) || len(p.ResourcesVpcConfig.PublicAccessCidrs) > 0 {
+		// Convert user-supplied slice of CIDRs to map of networks.
+		netMap := map[string]bool{}
+		for _, c := range p.ResourcesVpcConfig.PublicAccessCidrs {
+			_, net, err := net.ParseCIDR(c)
+			if err != nil {
+				return false, err
+			}
+			netMap[net.String()] = true
+		}
+		// If length of networks does not match the length of CIDR blocks
+		// returned by AWS then we need update.
+		if len(netMap) != len(cluster.ResourcesVpcConfig.PublicAccessCidrs) {
+			return false, nil
+		}
+		// If AWS returns a CIDR block that is not in the map, then we need
+		// update.
+		for _, pc := range cluster.ResourcesVpcConfig.PublicAccessCidrs {
+			if !netMap[pc] {
+				return false, nil
+			}
 		}
 	}
 
-	return result
+	return cmp.Equal(&v1beta1.ClusterParameters{}, patch, cmpopts.EquateEmpty(),
+		cmpopts.IgnoreTypes(&v1alpha1.Reference{}, &v1alpha1.Selector{}),
+		cmpopts.IgnoreFields(v1beta1.ClusterParameters{}, "Tags"),
```

Comment:
This was another place I was looking at `RDSInstance`.
https://github.com/crossplane/provider-aws/blob/3238aeef3ccc0dab5672482577749432d2d60082/pkg/clients/rds/rds.go#L448
The result here is that `Tags` never get updated unless something else changes in the `Spec`, which is not desirable. However, I believe the reason this is happening is because there will always be a diff with tags because of the way late initialization is ignoring them (per my earlier comment). We could probably fix this in both places by just doing a totally separate check on the tags. What do you think?

Reply:
@muvaf I have updated how this is handled in this commit: https://github.com/crossplane/provider-aws/pull/245/commits/6359cf52b4f23762e6dc37e9d40d9d1121889f0c